### PR TITLE
[DRAFT] Gesture navigation API + swipe implementation

### PR DESCRIPTION
**Pull request by KovalevAndrey**
## Description

Basic gesture API + Swipe implementation. Implemented as a Modifier which is applied to container composable.

If we agree on the API more gestures to come:
- draggable  
- multi touching 
- ..

Fixes https://github.com/bumble-tech/appyx/issues/97

https://user-images.githubusercontent.com/5773436/186649066-32f5b004-6ce2-4587-bca2-84000399fff7.mp4

## Check list

- [ ] I have updated `CHANGELOG.md` if required.
- [ ] I have updated documentation if required.

### CODE REVIEW

File: core/src/main/kotlin/com/bumble/appyx/navmodel/spotlight/gesturehandler/SpotlightSwipeGestureHandler.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import com.bumble.appyx.core.navigation.gesture.GestureHandler
+import com.bumble.appyx.core.navigation.gesture.swipe.SwipeGesture
+import com.bumble.appyx.navmodel.spotlight.Spotlight
+import com.bumble.appyx.navmodel.spotlight.operation.next
+import com.bumble.appyx.navmodel.spotlight.operation.previous
+
+class SpotlightSwipeGestureHandler<Routing : Any>(
+    val spotlight: Spotlight<Routing>
```
Comment (KovalevAndrey) [inline core/src/main/kotlin/com/bumble/appyx/navmodel/spotlight/gesturehandler/SpotlightSwipeGestureHandler.kt:12]: private

**Comment by LachlanMcKee** [comment]
Might be worth adding a video (when the PR is ready) to show how it works (with touches displayed)

File: core/src/main/kotlin/com/bumble/appyx/core/navigation/gesture/swipe/SwipeGestureModifierFactory.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            ) { _, dragAmount ->
+                val (x, y) = dragAmount
+                offsetX += x
+                offsetY += y
+            }
+        }
+    }
+}
+
+private const val SwipeThreshold = 200f
```
[the remark this answers is not in the document]
Comment (KovalevAndrey) [reply core/src/main/kotlin/com/bumble/appyx/core/navigation/gesture/swipe/SwipeGestureModifierFactory.kt:51]: Conflicts with https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md#singletons_constants_sealed-class-and-enum-class-values

**Comment by CherryPerry** [comment]
Are you sure that is what we need?

I imagined an implementation like in Telegram where a screen follows a finger. It would need to change `Children/Child` implementations to render not only top active item, but also top non-active.

**Comment by KovalevAndrey** [comment]
> Are you sure that is what we need?
> 
> I imagined an implementation like in Telegram where a screen follows a finger. It would need to change `Children/Child` implementations to render not only top active item, but also top non-active.

We can discuss what we need but I can imaging it's a different API. The problem with this is that in`Children/Child` you don't know in which states you children can be and this Telegram style navigation can be applied only for specific nav models like BackStack

File: core/src/main/kotlin/com/bumble/appyx/core/navigation/gesture/swipe/SwipeGestureModifierFactory.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (LachlanMcKee) [reply core/src/main/kotlin/com/bumble/appyx/core/navigation/gesture/swipe/SwipeGestureModifierFactory.kt:51]: Here is the change that should be made:
https://detekt.dev/docs/introduction/compose/#toplevelpropertynaming

**Comment by zsoltk** [comment]
@KovalevAndrey the gestures would be paired to a specific `NavModel` just like a `TransitionHandler` is, I don't see an issue with that (e.g. having horizontal swiping in `Tiles` wouldn't make too much sense). I'm also for the instant response to the gesture. 

I think we could:

1. Use the swipeable api
2. Anchors would trigger routing changes
3. For an instant animation of a target we would need to apply the same `Modifier` as we do in the corresponding `TransitionHandler`.

Possibly we could make all `TransitionHandlers` expect to work with a `Float`-based transition spec (as in 0..1f progress), and expose the part which maps `Float` → `Modifier`. That way we could probably reuse it in the gesture handling.

**Pull request closed without merging**